yt/frontends/stream/data_structures.py

```diff
@@ -980,36 +995,38 @@ def refine_amr(base_ds, refinement_criteria, fluid_operators, max_level,
                 LE = sg.left_index * g.dds + ds.domain_left_edge
                 dims = sg.dimensions * ds.refine_by
                 grid = ds.smoothed_covering_grid(g.Level + 1, LE, dims)
-                gd = dict(left_edge = LE, right_edge = grid.right_edge,
-                          level = g.Level + 1, dimensions = dims)
+                gd = dict(left_edge=LE, right_edge=grid.right_edge,
+                          level=g.Level + 1, dimensions=dims)
                 for field in ds.field_list:
                     if not isinstance(field, tuple):
                         field = ("unknown", field)
                     fi = ds._get_field_info(*field)
-                    if not fi.particle_type :
+                    if not fi.particle_type:
                         gd[field] = grid[field]
                 grid_data.append(gd)
 
         ds = load_amr_grids(grid_data, ds.domain_dimensions, bbox=bbox)
 
-        if number_of_particles > 0:
-            if ("io", "particle_position_x") not in pdata:
-                pdata_ftype = {}
-                for f in [k for k in sorted(pdata)]:
-                    if not hasattr(pdata[f], "shape"): continue
-                    mylog.debug("Reassigning '%s' to ('io','%s')", f, f)
-                    pdata_ftype["io",f] = pdata.pop(f)
-                pdata_ftype.update(pdata)
-                pdata = pdata_ftype
-            assign_particle_data(ds, pdata)
-            # We need to reassign the field list here.
         cur_gc = ds.index.num_grids
 
+    ds.particle_types_raw = base_ds.particle_types_raw
+    ds.particle_types = ds.particle_types_raw
+
+    # Now figure out where the particles go
+    if number_of_particles > 0:
+        # This will update the stream handler too
+        assign_particle_data(ds, pdata)
+        # Because we've already used the index, we
+        # have to re-create the field info because
+        # we added particle data after the fact
```

Comment:
I'm open to suggestions, but I'm not sure how. The problem is that in order to do the refinement we've already generated the `index`, but without the particle fields inside.